### PR TITLE
Support multiple marker types in WebGL scatter

### PR DESCRIPTION
**Pull request by ianthomas23**
Previously if you used the WebGL backend and called `scatter` with more than one marker type, e.g.
```python
p = figure(output_backend="webgl", ...)
p.scatter(marker=["circle", "square"], ...)
```
then rendering would drop back to using canvas. This PR addresses this, fixing item 4 of discussion #11369.

Changes:
  1. `MarkerGL` used to accept the `marker_type` in its constructor. This is no longer required.
  2. `MarkerGL` identifies the unique marker types it needs to render, and calls the ReGL rendering function once for each type. All but one array/buffer is common in these calls so the data transfer to GPU is minimal, only the `show` array/buffer is different each time.
  3. `ScatterView` no longer checks for the type of marker to change the underlying `glglyph`, it keeps the same `glglyph`.
  4. `ScatterView` assumes that `MarkerGL` can render all marker types, so I have removed the `is_supported` function. If a new marker type is added for canvas it must also be added for WebGL. This may need discussion.
  5. I have added a new visual test containing multiple marker types in a single scatter call. This has always worked of course, but previously the WebGL rendering dropped back to canvas. Image before this PR:
before
and after
after
The WebGL-trained eyeball will note that in the before image the webgl output is the same as canvas, but it is different in the after image as the antialiasing is different for webgl: the cross is darker and the dots are lighter for example. Any future regression of webgl back to canvas will be caught by this test.

I have kept the handling of both `circle` and `scatter` markers in `MarkerGL` even though it is a bit clunky (it always was, and dates back to "before my time" here). I intend to refactor this when I combine `circle`, `scatter` and `rect` into a single united implementation, presumably through a more sensible class hierarchy in the webgl code.

### CODE REVIEW

File: bokehjs/src/lib/models/glyphs/glyph.ts [hunk 1]
```diff
@@ -255,13 +255,9 @@ export abstract class GlyphView extends View {
       visual.update()
     }
 
-    this._set_visuals()
-
     this.glglyph?.set_visuals_changed()
   }
 
-  protected _set_visuals(): void {}
-
```
Comment (ianthomas23) [inline bokehjs/src/lib/models/glyphs/glyph.ts:264]: This is undoing the changes I introduced in PR #11297 as they are no longer required.

File: bokehjs/src/lib/models/glyphs/webgl/markers.ts [hunk 2]
```diff
@@ -107,42 +72,58 @@ export class MarkerGL extends BaseGLGlyph {
...
+
+      let nshow = nmarkers  // Number of markers to show.
+      const prev_nmarkers = this._show.length
+      const show_array = this._show.get_sized_array(nmarkers)
+      if (ntypes > 1 || indices.length < nmarkers) {
+        this._show_all = false
+
+        // Reset all show values to zero.
+        for (let i = 0; i < nmarkers; i++)
+          show_array[i] = 0
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/markers.ts:89]:
```suggestion
show_array.fill(0)
```

File: bokehjs/src/lib/models/glyphs/webgl/markers.ts [hunk 3]
```diff
@@ -107,42 +72,58 @@ export class MarkerGL extends BaseGLGlyph {
...
+
+        // Set show values of markers to render to 255.
+        nshow = 0
+        for (let j = 0; j < indices.length; j++) {
+          const k = indices[j]  // Marker index.
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/markers.ts:94]: Unless you need `j` for other purpose, it's simpler and faster to use this:
```suggestion
        for (const k of indices) {
```

File: bokehjs/src/lib/models/glyphs/webgl/markers.ts [hunk 4]
```diff
@@ -107,42 +72,58 @@ export class MarkerGL extends BaseGLGlyph {
...
+          if (ntypes == 1 || mainGlGlyph._marker_types!.get(k) == marker_type) {
+            show_array[k] = 255
+            nshow++
+          }
+        }
+      } else if (!this._show_all || prev_nmarkers != nmarkers) {
+        this._show_all = true
+        for (let i = 0; i < nmarkers; i++)
+          show_array[i] = 255
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/markers.ts:103]:
```suggestion
        show_array.fill(255)
```

File: bokehjs/test/integration/glyphs/markers.ts [hunk 5]
```diff
@@ -97,4 +90,25 @@ describe("Marker glyph", () => {
 
     await display(row([p0, p1, p2]))
   })
+
+  it("should support multiple marker types in scatter", async () => {
+    const marker_type = [
+      "asterisk", "circle", "square_dot", "triangle_pin", "star",
+      "x", "diamond_cross", "star", "circle_cross", "hex_dot"]
```
Comment (mattpap) [inline bokehjs/test/integration/glyphs/markers.ts:97]: Preferably use a trailing comma and move `]` to a new line. It's easier to modify an array this way.

File: bokehjs/src/lib/models/glyphs/webgl/markers.ts [hunk 6]
```diff
@@ -107,42 +72,58 @@ export class MarkerGL extends BaseGLGlyph {
...
-      // Reset all show values to zero.
-      for (let i = 0; i < nmarkers; i++)
-        show_array[i] = 0
-
-      // Set show values of markers to render to 255.
-      for (let j = 0; j < indices.length; j++) {
-        show_array[indices[j]] = 255
+    const ntypes = mainGlGlyph._unique_marker_types.length
+    for (let i = 0; i < ntypes; i++) {
+      const marker_type = mainGlGlyph._unique_marker_types[i]
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/markers.ts:77]:
```suggestion
    for (const marker_type of mainGlGlyph._unique_marker_types) {
```

**Comment by ianthomas23** [comment]
Argh, I forgot to set the labels yet again :astonished:

**Comment by bryevdv** [comment]
@ianthomas23 is this `ready` ?

**Comment by mattpap** [comment]
@ianthomas23, does this approach preserve z-order of consecutive data points?

**Comment by bryevdv** [comment]
> @ianthomas23, does this approach preserve z-order of consecutive data points?

I don't see how it could, webgl naturally wants to draw the same things in batches. Assuming that is so, I am definitely in favor of simply documenting that trade-off and letting users make the decision that fits their needs accordingly.

**Comment by ianthomas23** [comment]
> > @ianthomas23, does this approach preserve z-order of consecutive data points?

Not always.

> I don't see how it could, webgl naturally wants to draw the same things in batches.

Correct.

The marker WebGL shaders can only render a single marker type at a time. The rendering of a `scatter` containing multiple `marker` types will call the WebGL renderer once for each unique marker type, and the order of these is the same as the order that the marker types are first mentioned in the `marker` property. So if each marker type is contiguous in the `marker` property then the rendering will have the same order as in canvas. But if a marker type is not contiguous, e.g. `marker=["star", "circle", "star"]` then the order will be different as all stars will be rendered before all circles.

It doesn't have to be this way. It would not be difficult to rewrite the code to render each group of contiguous marker types in order. But it is easy to invent some pathological cases for which the performance would be atrocious, e.g. a single `scatter` call of 100,000 markers which alternate between `"star"` and `"circle"` would require 100,000 ReGL rendering calls with this approach rather than the 2 rendering calls of this PR.

**Comment by bryevdv** [comment]
@ianthomas23 can you add a note to the `scatter` and `Scatter` docs explaining that the draw order may vary? I don't think we need to go into detail, or in to why, just enough to let users know this is an explicit trade-off for the better performance of webgl. cc @tcmetzger for any docs input

**Comment by ianthomas23** [comment]
> @ianthomas23 can you add a note to the `scatter` and `Scatter` docs explaining that the draw order may vary? I don't think we need to go into details of why, just enough to let users know this is an explicit trade-off that exists for the better performance of webgl. cc @tcmetzger for any docs input

Yes, I will do that tomorrow.

**Comment by mattpap** [comment]
I would prefer all backends to have principally the same semantics, so if we allow grouped renderer of scatter markers in webgl, then we should allow that in canvas and svg as well. Though this is a topic for a broader discussion; for the time being, this PR is fine.

**Pull request merged**